jupyterhub/events.py

```diff
@@ -0,0 +1,133 @@
+"""
+Emit structured, discrete events when various actions happen.
+"""
+from traitlets.config import Configurable
+
+import logging
+from datetime import datetime
+import jsonschema
+from pythonjsonlogger import jsonlogger
+from traitlets import TraitType, List
+import json
+import six
+
+
+class Callable(TraitType):
+    """
+    A trait which is callable.
+
+    Classes are callable, as are instances
+    with a __call__() method.
+    """
+    info_text = 'a callable'
+    def validate(self, obj, value):
+        if six.callable(value):
+            return value
+        else:
+            self.error(obj, value)
+
+def _skip_message(record, **kwargs):
+    """
+    Remove 'message' from log record.
+
+    It is always emitted with 'null', and we do not want it,
+    since we are always emitting events only
+    """
+    del record['message']
+    return json.dumps(record, **kwargs)
+
+
+class EventLog(Configurable):
+    """
+    Send structured events to a logging sink
+    """
+    handlers_maker = Callable(
+        None,
+        config=True,
+        allow_none=True,
+        help="""
+        Callable that returns a list of logging.Handler instances to send events to.
+
+        When set to None (the default), events are discarded.
+        """
+    )
+
+    allowed_schemas = List(
+        [],
+        config=True,
+        help="""
+        Fully qualified names of schemas to record.
+
+        Each schema you want to record must be manually specified.
+        The default, an empty list, means no events are recorded.
+        """
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.log = logging.getLogger(__name__)
+        # We don't want events to show up in the default logs
+        self.log.propagate = False
+        self.log.setLevel(logging.INFO)
```

Comment:
I'm curious. Does log-level have any meaning for the event logger? Or is this just to make the default calls to `.emit()` pass through?

Reply:
It's to make the default calls to .emit pass through.